### PR TITLE
feat(.tool-versions): support for a new tool

### PR DESCRIPTION
**Pull request by v1v**
Support for `.tool-versions` format, see https://asdf-vm.com/manage/configuration.html#tool-versions

## Test

To test this pull request, you can run the following commands:

```shell
cd pkg/plugins/resources/toolversions 
go test

PASS
ok  	github.com/updatecli/updatecli/pkg/plugins/resources/toolversions	1.030s
```

```shell
cd pkg/plugins/utils/toolversions
go test
PASS
ok  	github.com/updatecli/updatecli/pkg/plugins/utils/toolversions	0.378s
```

### Potential improvement

Support for the `autodiscovery`?

### CODE REVIEW

File: e2e/updatecli.d/success.d/toolversions.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+name: Basic .tool-versions Example
```
Comment (v1v) [inline e2e/updatecli.d/success.d/toolversions.yaml:1]: I'm trying to test this out now:

```bash
dist/updatecli_darwin_arm64/updatecli --debug diff --config e2e/updatecli.d/success.d/toolversions.yaml
```

but it failed with:

```

################################
# BASIC .TOOL-VERSIONS EXAMPLE #
################################


SOURCES
=======

local
-----
ERROR: ✗ ⚠ Don't support resource kind: toolversions
Pipeline "Basic .tool-versions Example" failed
Skipping due to:
	sources stage:	"⚠ Don't support resource kind: toolversions"

```

Comment (olblak) [reply e2e/updatecli.d/success.d/toolversions.yaml:1]: You also need to add toolversions to [GetResourceMapping](https://github.com/updatecli/updatecli/blob/f156b7b836b37489bb32b8eccbe1f8ae1c72e4e3/pkg/core/pipeline/resource/main.go#L239)

**Comment by olblak** [comment]
Is there an API that we can query to retrieve the package version?

**Comment by v1v** [comment]
> Is there an API that we can query to retrieve the package version?

I've been thinking of using an ad hoc implementation for some of the known tools, such as Golang, Python, Java, or similar ones.

For instance, there are some GitHub actions in charge of configuring some tools and installing the given version or aliases:
- https://github.com/actions/setup-node
- https://github.com/actions/setup-go
- https://github.com/actions/setup-python

I don't know if a similar implementation to what Jenkins does for crawling the tools:
- https://github.com/jenkins-infra/crawler

Could be an option?

What do you think? Am I thinking this too much, and it might not be something worthy to do it?

File: e2e/updatecli.d/success.d/toolversions.yaml (hunk 1 above)
Comment (v1v) [reply e2e/updatecli.d/success.d/toolversions.yaml:1]: Thanks! It works like a charm now:

```bash
 dist/updatecli_darwin_arm64/updatecli --debug diff --config e2e/updatecli.d/success.d/toolversions.yaml


+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "e2e/updatecli.d/success.d/toolversions.yaml"
DEBUG: pipelineid undefined, we'll try to generate one
DEBUG: using pipeline name to generate the pipelineid

SCM repository retrieved: 0


++++++++++++++++++
+ AUTO DISCOVERY +
++++++++++++++++++



++++++++++++
+ PIPELINE +
++++++++++++



################################
# BASIC .TOOL-VERSIONS EXAMPLE #
################################


SOURCES
=======

local
-----
✔ value "1.0.0", found in file "/Users/vmartinez/workspaces/elastic/updatecli/pkg/plugins/resources/toolversions/testdata/.tool-versions", for key "bats"'
DEBUG: empty changelog found for the source


CONDITIONS:
===========

local
-----
✔ key "bats", from file "pkg/plugins/resources/toolversions/testdata/.tool-versions", is correctly set to "1.0.0"

golang
------
✔ key "golang", from file "pkg/plugins/resources/toolversions/testdata/.tool-versions", is correctly set to "1.8.2"


TARGETS
========

local-source
------------

**Dry Run enabled**

⚠ - key "tool", from file "pkg/plugins/resources/toolversions/testdata/.tool-versions", is incorrectly set to "" and should be "1.0.0"

local-value
-----------

**Dry Run enabled**

⚠ - key "new-tool", from file "pkg/plugins/resources/toolversions/testdata/.tool-versions", is incorrectly set to "" and should be "2.0.0"


ACTIONS
========


=============================

REPORTS:



⚠ Basic .tool-versions Example:
	Source:
		✔ [local] Get value from .tool-versions
	Condition:
		✔ [golang] Test value from .tool-versions
		✔ [local] Test value from .tool-versions with sourceid
	Target:
		⚠ [local-source] Test sourceid new key from .tool-versions
		⚠ [local-value] Test new tool with value from .tool-versions


Run Summary
===========
Pipeline(s) run:
  * Changed:	1
  * Failed:	0
  * Skipped:	0
  * Succeeded:	0
  * Total:	1
```

**Comment by dduportal** [comment]
> > Is there an API that we can query to retrieve the package version?
> 
> I've been thinking of using an ad hoc implementation for some of the known tools, such as Golang, Python, Java, or similar ones.
> 
> For instance, there are some GitHub actions in charge of configuring some tools and installing the given version or aliases:
> 
>     * https://github.com/actions/setup-node
> 
>     * https://github.com/actions/setup-go
> 
>     * https://github.com/actions/setup-python
> 
> 
> I don't know if a similar implementation to what Jenkins does for crawling the tools:
> 
>     * https://github.com/jenkins-infra/crawler
> 
> 
> Could be an option?
> 
> What do you think? Am I thinking this too much, and it might not be something worthy to do it?

I believe an issue with discussion would be needed: ASDF support would be really cool, but we have to plan it all together as there are no centralized package index: it uses plugins for each implementation.

Plugin could be used to retrieve informations but it might be a case by case implementation

**Comment by v1v** [comment]
> I believe an issue with discussion would be needed: ASDF support would be really cool, but we have to plan it all together as there are no centralized package index: it uses plugins for each implementation.

See https://github.com/updatecli/updatecli/issues/2274

**Comment by olblak** [comment]
@v1v Could I also ask you to create a documentation page on https://github.com/updatecli/website/tree/master/content/en/docs/plugins/resource ? 

I see a few spelling issues with unknown words, you can add them to "https://github.com/updatecli/updatecli/blob/main/.github/actions/spelling/allow.txt" to ignore those message

Feel free to ping if you need some help to finalize this pullrequest

File: pkg/plugins/resources/toolversions/main.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+var (
+	// ErrDaselFailedParsingToolsVersionsByteFormat is returned if dasel couldn't parse the byteData
+	ErrDaselFailedParsingToolsVersionsByteFormat error = errors.New("fail to parse .tool-versions data")
+)
+
+// ToolVersions stores configuration about the file and the key value which needs to be updated.
+type ToolVersions struct {
+	spec     Spec
+	contents []toolversions.FileContent
```
Comment (v1v) [inline pkg/plugins/resources/toolversions/main.go:20]: Do I need [`505b162` (#2265)](https://github.com/updatecli/updatecli/pull/2265/commits/505b162e626871f49120ba94b419b97140c49f85)?

**Comment by v1v** [comment]
> @v1v Could I also ask you to create a documentation page on [updatecli/website@master/content/en/docs/plugins/resource](https://github.com/updatecli/website/tree/master/content/en/docs/plugins/resource?rgh-link-date=2024-07-05T07%3A04%3A42Z) ?

Done, see https://github.com/updatecli/website/pull/1477

File: pkg/plugins/resources/toolversions/main.go (hunk 2 above)
Comment (olblak) [reply pkg/plugins/resources/toolversions/main.go:20]: Do you mean `foundVersion version.Version`?
I don't think so, as far as I understand this plugin, you don't need to filter version based on rules like semantic versioning ,etc.

**Comment by olblak** [comment]
@v1v Are you confident merging this pullrquest? 
I did a quick test with your e2e tests and it sounds great

File: pkg/plugins/resources/toolversions/main.go (hunk 2 above)
Comment (v1v) [reply pkg/plugins/resources/toolversions/main.go:20]: > Do you mean foundVersion version.Version?

Yes. Thanks for the clarification :)

**Comment by v1v** [comment]
> @v1v Are you confident merging this pullrquest?

Yes.

I ran a few local executions with different cases and so far it works as I intended 🙏

**Pull request merged**